modules/currency-exchange/src/lib.rs

```diff
@@ -0,0 +1,403 @@
+	type CurrencyConverter: CurrencyConverter<
+		SourceAmount = <Self::PeerMaybeLockFundsTransaction as MaybeLockFundsTransaction>::Amount,
+		TargetAmount = Self::Amount,
+	>;
+	/// Something that could grant money.
+	type Airdrop: Airdrop<Recipient = Self::AccountId, Amount = Self::Amount>;
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// Invalid peer blockchain transaction provided.
+		InvalidTransaction,
+		/// Peer transaction has invalid amount.
+		InvalidAmount,
+		/// Peer transaction has invalid recipient.
+		InvalidRecipient,
+		/// Cannot map from peer recipient to this blockchain recipient.
+		FailedToMapRecipients,
+		/// Failed to convert from peer blockchain currency to this blockhain currency.
+		FailedToCovertCurrency,
+		/// Airdrop has failed.
+		AirdropFailed,
+		/// Transaction is not finalized.
+		UnfinalizedTransaction,
+		/// Transaction funds are already claimed.
+		AlreadyClaimed,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Imports lock fund transaction of the peer blockchain.
+		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
+		pub fn import_peer_transaction(
+			origin,
+			transaction: <<T as Trait>::PeerBlockchain as Blockchain>::Transaction,
+			block: <<T as Trait>::PeerBlockchain as Blockchain>::BlockHash,
+			proof: <<T as Trait>::PeerBlockchain as Blockchain>::TransactionInclusionProof,
+		) -> DispatchResult {
+			frame_system::ensure_signed(origin)?;
+
+			// ensure that transaction is included in finalized block that we know of
+			//
+			// note that the transaction itself is actually redundant here, because it
+			// will probably be a part of proof itself (like it is now), or could be
+			// reconstructed from proof during verification
+			//
+			// leaving it here for now just for simplicity
+			let is_transaction_finalized = <T as Trait>::PeerBlockchain::verify_transaction_inclusion_proof(
+				&transaction,
+				block,
+				&proof,
+			);
+			if !is_transaction_finalized {
+				fail!(Error::<T>::UnfinalizedTransaction);
+			}
+
+			// parse transaction
+			let transaction = <T as Trait>::PeerMaybeLockFundsTransaction::parse(&transaction)
+				.map_err(Error::<T>::from)?;
+			let transfer_id = transaction.id.encode();
+			ensure!(
+				!Transfers::contains_key(&transfer_id),
+				Error::<T>::AlreadyClaimed
+			);
+
+			// grant recipient
+			let recipient = T::RecipientsMap::map(transaction.recipient).map_err(Error::<T>::from)?;
+			let amount = T::CurrencyConverter::convert(transaction.amount).map_err(Error::<T>::from)?;
+			T::Airdrop::drop(recipient, amount).map_err(Error::<T>::from)?;
+
+			// remember that we have accepted this transfer
+			Transfers::insert(transfer_id, ());
+
+			Ok(())
+		}
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Bridge {
+		/// All transfers that have already been claimed.
+		Transfers: map hasher(blake2_128_concat) Vec<u8> => ();
```

Comment:
Why `Vec<u8>` and not the `EthTransactionTag` stronger type? The result would be the same anyway, wouldn't it? But at least there would be some reference to the key format.